programs/zstdcli.c

```diff
@@ -346,9 +346,22 @@ static int readU32FromCharChecked(const char** stringPtr, unsigned* value)
         (*stringPtr)++;  /* skip `K` or `M` */
         if (**stringPtr=='i') (*stringPtr)++;
         if (**stringPtr=='B') (*stringPtr)++;
+
+    }
+
+    //Only well formed entries will be 0, or , (specifically for zstd compound)
+    if (**stringPtr == 0 || **stringPtr == ',') {
+
+        *value = result;
+        return 0;
+
+    } else {
+
+        *value = 0;
+        return 2;
+
```

Comment:
We can use this opportunity to reduce code duplication in this file.
Instead of implementing the same logic in both `readU32FromCharChecked` and `readSizeTFromCharChecked` we can use the fact that `size_t` will always be at least as big a `U32`.
Then we can have `readU32FromCharChecked` call `readSizeTFromCharChecked` and just check for U32 overflow.

Reply:
@yoniko - interesting point - can we assume that there aren't any 16 bit processors knocking around which will use this software (I'd assume something _very_ esoteric!). I did mull merging these - what are the 2 use cases that led to there being 2 functions? Both are unsigned integers - but one is (possibly) clamped to smaller than word size?

Reply:
Yes, one supports up to 32bit while the other supports up to size_t (32 or 64 bit).

Looks like the size_t function was added at https://github.com/facebook/zstd/commit/834d06b6 and was probably due to someone wanting to assign sizes that are larger than 4GB. I don't really like this previous patch and I think it creates some inconsistencies in the cli, but we shouldn't address those now.

You can assume size_t is at least 32 bit (I believe the codebase wouldn't work if size_t is 16 bit), but you can add a static_assert just in case.